src/common/cups_print.c

```diff
@@ -0,0 +1,538 @@
+/*
+    This file is part of darktable,
+    copyright (c) 2014-2015 pascal obry.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <glib.h>
+#include <stdio.h>
+#include <cups/cups.h>
+#include <cups/ppd.h>
+
+#include "common/image.h"
+#include "common/image_cache.h"
+#include "common/mipmap_cache.h"
+#include "common/pdf.h"
+#include "cups_print.h"
+#include "control/jobs/control_jobs.h"
+
+// initialize the pinfo structure
+void dt_init_print_info(dt_print_info_t *pinfo)
+{
+  memset(&pinfo->printer, 0, sizeof(dt_printer_info_t));
+  memset(&pinfo->page, 0, sizeof(dt_page_setup_t));
+  memset(&pinfo->paper, 0, sizeof(dt_paper_info_t));
+  pinfo->printer.intent = DT_INTENT_PERCEPTUAL;
+  *pinfo->printer.profile = '\0';
+}
+
+dt_printer_info_t *dt_get_printer_info(const char *printer_name)
+{
+  cups_dest_t *dests;
+  int num_dests = cupsGetDests(&dests);
+  cups_dest_t *dest = cupsGetDest(printer_name, NULL, num_dests, dests);
+  dt_printer_info_t *result = NULL;
+
+  if (dest)
+  {
+    const char *PPDFile = cupsGetPPD (printer_name);
+    result = (dt_printer_info_t *)malloc(sizeof(dt_printer_info_t));
+    g_strlcpy(result->name, dest->name, MAX_NAME);
+    ppd_file_t *ppd = ppdOpenFile(PPDFile);
+
+    if (ppd)
+    {
+      ppdMarkDefaults(ppd);
+      cupsMarkOptions(ppd, dest->num_options, dest->options);
+
+      // hardware margins
+
+      ppd_attr_t *attr = ppdFindAttr(ppd, "HWMargins", NULL);
+
+      if (attr)
+      {
+        sscanf(attr->value, "%lf %lf %lf %lf",
+               &result->hw_margin_left, &result->hw_margin_bottom,
+               &result->hw_margin_right, &result->hw_margin_top);
+
+        result->hw_margin_left   = dt_pdf_point_to_mm (result->hw_margin_left);
```

Comment:
I don't think that you should be using dt_pdf_point_to_mm here, or is CUPS reporting the margings in PDF points (1/72 inch)?